packages/zql-integration-tests/src/helpers/runner.ts

```diff
@@ -304,49 +317,188 @@ export async function createVitests<TSchema extends Schema>(
+  }[])[]
+): Promise<void>;
+export async function runBenchmarks<TSchema extends Schema>(
+  {suiteName, type, zqlSchema, pgContent, only}: BenchOptions<TSchema>,
+  ...benchSpecs: (readonly {
+    name: string;
+    createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
+    generatePush?: PushGenerator;
+  }[])[]
+): Promise<void> {
   const dbs = await makeDatabases(suiteName, zqlSchema, pgContent);
   const delegates = makeDelegates(dbs, zqlSchema);
   const queries = makeQueries(zqlSchema, delegates);
 
-  return benchSpecs
+  benchSpecs
     .flat()
     .filter(t => (only ? only.includes(t.name) : true))
-    .map(({name, createQuery}) =>
+    .map(({name, createQuery, generatePush}) =>
       describe(name, () => {
-        makeHydrationBenchmark({
+        makeBenchmark({
+          zqlSchema,
+          delegates,
+          dbs,
           createQuery,
+          type,
           queries,
+          generatePush,
         });
       }),
     );
 }
 
-function makeHydrationBenchmark<TSchema extends Schema>({
+function makeBenchmark<TSchema extends Schema>({
+  zqlSchema,
   createQuery,
+  type,
   queries,
+  delegates,
+  dbs,
+  generatePush,
 }: {
+  zqlSchema: TSchema;
   createQuery: (q: Queries<TSchema>) => Query<TSchema, string>;
+  type: 'hydration' | 'push';
   queries: QueriesBySource<TSchema>;
+  delegates: Delegates;
+  dbs: DBs<TSchema>;
+  generatePush: PushGenerator | undefined;
 }) {
-  const zql = createQuery(queries.memory);
-  const zqlite = createQuery(queries.sqlite);
-  const pg = createQuery(queries.pg);
-  return [doBench('zql', zql), doBench('zqlite', zqlite), doBench('zpg', pg)];
+  switch (type) {
+    case 'push': {
+      // isolate delegates to own transactions
+      // so push benchmarks are isolated from each other.
+      const newDelegates = {
+        ...delegates,
+        memory: new TestMemoryQueryDelegate({
+          sources: Object.fromEntries(
+            Object.entries(dbs.memory).map(([key, source]) => [
+              key,
+              source.fork(),
+            ]),
+          ),
+        }),
+        sqlite: newQueryDelegate(
+          lc,
+          testLogConfig,
+          (() => {
+            const db = new Database(lc, dbs.sqliteFile);
+            db.exec('BEGIN CONCURRENT');
+            return db;
+          })(),
+          zqlSchema,
+        ),
+      };
+      const queryBuilders = makeQueries(zqlSchema, newDelegates);
+      const zqlite = createQuery(queryBuilders.sqlite);
+      const zql = createQuery(queryBuilders.memory);
+
+      // materialize before starting the benchmark. We only want to time push
+      // and not the initial hydration.
+      const zqliteView = zqlite.materialize();
+      const zqlView = zql.materialize();
+      try {
+        benchPush('zql', newDelegates, must(generatePush));
+        benchPush('zqlite', newDelegates, must(generatePush));
+      } finally {
+        zqliteView.destroy();
+        zqlView.destroy();
```

Comment:
Are you sure that the benchmarks runs before this? I would have thought benchPush sets up the benchmark but does not actually run the benchmark.

Reply:
Yeah, you're probably right. I'll check.